### DISCUSSION TITLE
Auto-restarting frankenphp in worker mode on file changes (Docker local development)

### DISCUSSION THREAD

**Discussion by binaryfire** [php/frankenphp#556] [Q&A]
We've just migrated from the binary to the Docker image for local Octane development and need to implement a way to reload / restart frankenphp on file changes. Could someone point me in the right direction? Using non-worker mode in local dev isn't an option for us.

From what I can see, the `frankenphp reload` command just reloads the Caddy config. Octane handles hot reloads natively when using the binary. Not sure how to get the same thing happening in the container.

**Comment by withinboredom** [comment]
This is the script I use to do a few things on file changes -- but I don't use worker mode in development:

```sh
#!/bin/bash

sigint_handler()
{
  kill $PID
  exit
}

trap sigint_handler SIGINT

while true; do
  composer dump -o --apcu
  twcss -i public/assets/style.css -o public/assets/web.css --minify
  $@ &
  PID=$!
  inotifywait -e modify -e move -e create -e delete -e attrib -r "$(pwd)"
  kill $PID
done
```

This runs `composer dump` (I use `composer-attribute-collector` to scan for attributes), then tailwind CSS, then whatever I told it to do:

```
./watch frankenphp run
```

will watch the current directory for changes, and restart things.

I don't use a mac for development, so this likely needs some changes to support a mac, which uses a totally different filesystem/watching method.

**Reply by binaryfire** [reply]
Nice, thanks for sharing. I ended up adding Supervisor to the container and using a Chokidar file watcher to fire a reload command based on [Octane's reload logic](https://github.com/laravel/octane/blob/8f9d0da559ba4366fecc5c106d1298abe6773396/src/FrankenPhp/ServerProcessInspector.php#L39). You can just hit the admin endpoint to trigger a stop+start, which is handy.

We've got multiple devs in the team (all using different platforms) so I needed a way of self-containing the reload logic.

**Reply by dunglas** [reply]
I plan to add this feature natively in the next minor version.

**Reply by binaryfire** [reply]
Great! So it'll be a file watcher that monitors the `/app` dir and restarts the server on changes? I'm assuming it'll be an optional flag since we wouldn't want it running in certain environments.

**Reply by dunglas** [reply]
Yes!

**Reply by mateusfmello** [reply]
> I plan to add this feature natively in the next minor version.

Has hot loading already been implemented natively?

I'm using FrankenPHP for the first time and the changes are taking effect inside the Docker container.

**Comment by SanaMegabytes** [comment]
{
	{$CADDY_GLOBAL_OPTIONS}

    frankenphp {
        worker {
            file  /app/public/index.php
            watch
        }
    }
}

in Caddyfile !
